### PR TITLE
Fix testReadFailure2 using UnicodeBOMInputStream

### PR DESCRIPTION
**Pull request by antoine-guillou**

### CODE REVIEW

**Comment by benmccann** [comment]
@antoine-guillou can you rebase this PR against the latest master?

**Comment by antoine-guillou** [comment]
Ok, I updated my fork, solved the conflicts and pushed the branch.

Note that I only perform the UnicodeBOM check in the last **read** method, as all the other read methods end up in this one.
I also perform the check in **headerOnly** and **detectedColumnTypes**.

I am also wondering if we should close the streams at some point.

File: core/src/main/java/tech/tablesaw/io/csv/CsvReader.java [hunk 1]
```diff
@@ -336,7 +344,11 @@ public static Table headerOnly(ColumnType types[], boolean header, char columnSe
      */
     private static Table detectedColumnTypes(String csvFileName, boolean header, char delimiter) throws IOException {
         File file = new File(csvFileName);
-        ColumnType[] types = detectColumnTypes(new FileInputStream(file), header, delimiter, false);
+        InputStream fis = new FileInputStream(file);
```
Comment (benmccann) [inline core/src/main/java/tech/tablesaw/io/csv/CsvReader.java]: can you move this to happen inside `detectColumnTypes`?

**Comment by lwhite1** [comment]
AFAIK, the streams are generally closed either by opening them using
try-with-resources, or by closing the top level reader that reads them.

Do you see somewhere where that isn't the case?

File: core/src/main/java/tech/tablesaw/io/csv/CsvReader.java [hunk 2]
```diff
@@ -207,64 +207,69 @@ public static Table read(InputStream stream, String tableName, boolean header, c
...
+            String tableName,
+            ColumnType[] types,
+            boolean header,
+            char columnSeparator) throws IOException {
+
+        // All other read methods end up here, make sure we do not have leading Unicode BOM
+        UnicodeBOMInputStream ubis = new UnicodeBOMInputStream(stream);
+        ubis.skipBOM();
+
+        BufferedReader streamReader = new BufferedReader(new InputStreamReader(ubis));
```
Comment (benmccann) [inline core/src/main/java/tech/tablesaw/io/csv/CsvReader.java]: we shouldn't need to wrap the `InputStreamReader` in a `BufferedReader`. I would just inline this below:

    try (CSVReader reader = new CSVReaderBuilder(new InputStreamReader(ubis)).withCSVParser(csvParser).build()) {

**Comment by benmccann** [comment]
We close the `CSVReader` by including it in the `try`. The `CSVReader` closes the underlying `InputStream`, so we should be safe

File: core/src/test/java/tech/tablesaw/io/csv/CsvReaderTest.java [hunk 3]
```diff
@@ -2,19 +2,16 @@
...
-import java.net.URL;
-import java.util.Arrays;
-
 import org.junit.Ignore;
 import org.junit.Test;
-
 import tech.tablesaw.api.ColumnType;
 import tech.tablesaw.api.ShortColumn;
 import tech.tablesaw.api.Table;
+import java.io.File;
```
Comment (antoine-guillou) [inline core/src/test/java/tech/tablesaw/io/csv/CsvReaderTest.java]: ok, done

File: core/src/main/java/tech/tablesaw/io/csv/CsvReader.java (hunk 1 above)
Comment (antoine-guillou) [inline core/src/main/java/tech/tablesaw/io/csv/CsvReader.java]: ok, done

File: core/src/main/java/tech/tablesaw/io/csv/CsvReader.java (hunk 2 above)
Comment (antoine-guillou) [inline core/src/main/java/tech/tablesaw/io/csv/CsvReader.java]: corrected in the last push

**Comment by antoine-guillou** [comment]
Ok, I understand the closing strategy better now!

**Comment by benmccann** [comment]
Thanks for this!

**Comment by benmccann** [comment]
@antoine-guillou I just removed the old BOM workaround now that we have this one. You might want to do a pull and make sure it still works for you

**Comment by antoine-guillou** [comment]
Yes, everything is working fine on master. Cheers.

**Comment by lwhite1** [comment]
Thanks Antoine!

Ben, are you sure about the buffered reader?  I generally included them for
performance reasons.

**Comment by benmccann** [comment]
Yes, if you pass in a non-buffered reader, then it will create one for you:

        this.br =
                (reader instanceof BufferedReader ?
                        (BufferedReader) reader :
                        new BufferedReader(reader));

**Comment by lwhite1** [comment]
Cool. Thx

**Pull request merged**